### PR TITLE
Compact JSON-LD responses

### PR DESCRIPTION
**Pull request by thewilkybarkid**
Reduces the size of the responses by setting a default vocabulary and language in the context.

### CODE REVIEW

File: src/middleware/jsonld.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import jsonld from 'jsonld';
+import { Context as JsonLdContext } from 'jsonld/jsonld-spec';
```
Comment (thewilkybarkid) [inline src/middleware/jsonld.ts:3]: This is a TypeScript-only file, eslint can't find it so must be configured wrong.

File: src/middleware/jsonld.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+export default (context: JsonLdContext): Middleware => (
+  async ({ response }: Context, next: Next): Promise<void> => {
+    await next();
+
+    if (!isJsonLd(response)) {
+      return;
+    }
+
+    response.body = await jsonld.compact(response.body, context);
```
Comment (giorgiosironi) [inline src/middleware/jsonld.ts:23]: is compacting of universal utility or should it be skipped when using the development target (like minification and similar)?

Comment (thewilkybarkid) [reply src/middleware/jsonld.ts:23]: I think it's useful generally. It will be replaced with profile negotation at some point too.

File: src/middleware/jsonld.ts (hunk 1 above)
Comment (thewilkybarkid) [reply src/middleware/jsonld.ts:3]: https://github.com/libero/article-store/pull/17

File: src/app.ts [hunk 3]
```diff
@@ -13,6 +14,14 @@ app.use(logger());
...
   exposeHeaders: ['Link'],
 }));
+app.use(jsonld({
+  '@language': 'en',
+  '@vocab': 'http://schema.org/',
+  hydra: 'http://www.w3.org/ns/hydra/core#',
+  owl: 'http://www.w3.org/2002/07/owl#',
+  rdf: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
+  rdfs: 'http://www.w3.org/2000/01/rdf-schema#',
+}));
```
Comment (thewilkybarkid) [inline src/app.ts]: This isn't ideal, as introduces prefixes that might be unused, but the default might not also be appropriate (eg API documentation has no Schema.org).

File: test/middleware/jsonld.test.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+const makeRequest = async (next?: Next, jsonLdContext: JsonLdContext = {}): Promise<Response> => {
+  const context = createContext();
+
+  return runMiddleware(jsonld(jsonLdContext), context, next);
+};
+
+const next = (body: unknown, type: string) => async ({ response }: Context): Promise<void> => {
+  response.body = body;
+  response.type = type;
```
Comment (giorgiosironi) [inline test/middleware/jsonld.test.ts:14]: weird that it is the `next` middleware setting the body, although it checks that the `jsonld` middleware waits on it. If `response` was correctly populated at the start of the test this dummy middleware would be a simpler empty function?

File: test/middleware/jsonld.test.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    expect(contentType.parameters).toMatchObject({ profile: 'http://www.w3.org/ns/json-ld#compacted' });
+    expect(response.body).toMatchObject(expected);
+  });
+
+  it('should compact JSON-LD with a new context', async (): Promise<void> => {
+    const context = {
+      '@base': 'http://example.com',
+      '@language': 'en',
+      '@vocab': 'http://schema.org/',
+      dc: 'http://purl.org/dc/elements/1.1/',
```
Comment (giorgiosironi) [inline test/middleware/jsonld.test.ts:53]: inconsistent that these keys are unquoted, if they are meant to be used literally as `dc` or `name`?

File: test/middleware/jsonld.test.ts (hunk 4 above)
Comment (thewilkybarkid) [reply test/middleware/jsonld.test.ts:14]: Still need one case to make sure it runs after `next`.

File: test/middleware/jsonld.test.ts [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    const contentType = parseContentType(response);
+
+    expect(response.type).toBe('application/ld+json');
+    expect(contentType.type).toBe('application/ld+json');
+    expect(contentType.parameters).toMatchObject({});
+    expect(response.body).toBe('text');
+  });
+
+  it('should do nothing if the response JSON-LD but there is no body', async (): Promise<void> => {
+    const response = await makeRequest(next(undefined, 'jsonld'));
```
Comment (giorgiosironi) [inline test/middleware/jsonld.test.ts:85]: if there is no body how is the response JSON-LD or even JSON? What is the use case?

File: test/middleware/jsonld.test.ts (hunk 5 above)
Comment (thewilkybarkid) [reply test/middleware/jsonld.test.ts:53]: Yep, JS object and coding standard. (Doesn't help that a lot of the keys contain special characters.)

File: test/middleware/jsonld.test.ts (hunk 6 above)
Comment (thewilkybarkid) [reply test/middleware/jsonld.test.ts:85]: Edge case, and definitely shouldn't happen inside the app, but `Content-Type` headers can lie...

File: test/middleware/jsonld.test.ts (hunk 4 above)
Comment (giorgiosironi) [reply test/middleware/jsonld.test.ts:14]: Yes, but if that part breaks you get 1 failing test instead of all of them.

Comment (thewilkybarkid) [reply test/middleware/jsonld.test.ts:14]: Understand; think changing it doesn't make it simpler though. Happy if you'd like to refactor.

Comment (giorgiosironi) [reply test/middleware/jsonld.test.ts:14]: Not much value in changing towards either way for me, raised just for the sake of discussion.

**Review by giorgiosironi — APPROVED** [review]
Has conflicts to solve, but no further concerns
